src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/IIntervalTree`1.cs

```diff
@@ -0,0 +1,42 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.CodeAnalysis.Shared.Collections;
+
+/// <summary>
+/// Generic function representing the type of interval testing operation that can be performed on an interval tree. For
+/// example checking if an interval 'contains', 'intersects', or 'overlaps' with a requested span.
+/// </summary>
+internal delegate bool TestInterval<T, TIntrospector>(T value, int start, int length, in TIntrospector introspector)
+    where TIntrospector : struct, IIntervalIntrospector<T>;
+
+/// <summary>
+/// Base interface all interval trees need to implement to get full functionality.  Callers are not expected to use
+/// these methods directly.  Instead, they are the low level building blocks that the higher level extension methods are
+/// built upon. Consumers of an interface tree should use <c>.Algorithms</c> on the instance to get access to a wealth
+/// of fast operations through the <see cref="IntervalTreeAlgorithms{T, TIntervalTree}"/> type.
+/// </summary>
+internal interface IIntervalTree<T>
+{
+    /// <summary>
+    /// Adds all intervals within the tree within the given start/length pair that match the given <paramref
+    /// name="testInterval"/> predicate.  Results are added to the <paramref name="builder"/> array.  The <paramref
+    /// name="stopAfterFirst"/> indicates if the search should stop after the first interval is found.  Results will be
+    /// returned in a sorted order based on the start point of the interval.
+    /// </summary>
+    /// <returns>The number of matching intervals found by the method.</returns>
+    int FillWithIntervalsThatMatch<TIntrospector>(
+        int start, int length, TestInterval<T, TIntrospector> testInterval,
+        ref TemporaryArray<T> builder, in TIntrospector introspector,
+        bool stopAfterFirst)
+        where TIntrospector : struct, IIntervalIntrospector<T>;
+
+    /// <summary>
+    /// Practically equivalent to <see cref="FillWithIntervalsThatMatch{TIntrospector}"/> with a check that at least one
+    /// item was found.  However, separated out as a separate method as implementations can often be more efficient just
+    /// answering this question, versus the more complex "full with intervals" question above.
```

Comment:
>full [](http://example.com/codeflow?start=58&length=4)

nit: "fill with interfaces". I'll be interested in seeing how this differs from just calling fill to get the first one.